File: src/commands/grep.ts
/**
 * Grep command - Search tools by pattern
 */

import {
  type ToolInfo,
  type McpConnection,
  getConnection,
  debug,
  getConcurrencyLimit,
  safeClose,
} from '../client.js';
import {
  type McpServersConfig,
  getServerConfig,
  listServerNames,
  loadConfig,
} from '../config.js';
import { ErrorCode } from '../errors.js';
import { formatSearchResults } from '../output.js';

export interface GrepOptions {
  pattern: string;
  withDescriptions: boolean;
  configPath?: string;
}

interface SearchResult {
  server: string;
  tool: ToolInfo;
}

interface ServerSearchResult {
  serverName: string;
  results: SearchResult[];
  error?: string;
}

/**
 * Convert glob pattern to regex
 * Handles: * (any chars), ? (single char), ** (globstar)
 *
 * Examples:
 * - "*file*" matches "read_file", "file_utils"
 * - "**test**" matches "test", "my_test_tool", "testing"
 * - "server/*" matches "server/tool" but not "server/sub/tool"
 * - "server/**" matches "server/tool" and "server/sub/tool"
 */
export function globToRegex(pattern: string): RegExp {
  let escaped = '';
  let i = 0;

  while (i < pattern.length) {
    const char = pattern[i];

    if (char === '*' && pattern[i + 1] === '*') {
      // ** (globstar) - match anything including slashes (zero or more chars)
      escaped += '.*';
      i += 2;
      // Skip any immediately following * (e.g., *** becomes .*)
      while (pattern[i] === '*') {
        i++;
      }
    } else if (char === '*') {
      // * - match any chars except slash (zero or more)
      escaped += '[^/]*';
      i += 1;
    } else if (char === '?') {
      // ? - match single char (not slash)
      escaped += '[^/]';
      i += 1;
    } else if ('[.+^${}()|\\]'.includes(char)) {
      // Escape special regex chars
      escaped += `\\${char}`;
      i += 1;
    } else {
      escaped += char;
      i += 1;
    }
  }

  return new RegExp(`^${escaped}$`, 'i');
}

/**
 * Process items with limited concurrency, preserving order
 */
async function processWithConcurrency<T, R>(
  items: T[],
  processor: (item: T, index: number) => Promise<R>,
  maxConcurrency: number,
): Promise<R[]> {
  const results: R[] = new Array(items.length);
  let currentIndex = 0;

  async function worker(): Promise<void> {
    while (currentIndex < items.length) {
      const index = currentIndex++;
      results[index] = await processor(items[index], index);
    }
  }

  // Start workers up to concurrency limit
  const workers = Array.from(
    { length: Math.min(maxConcurrency, items.length) },
    () => worker(),
  );

  await Promise.all(workers);
  return results;
}

/**
 * Search tools in a single server (uses daemon if enabled)
 */
async function searchServerTools(
  serverName: string,
  config: McpServersConfig,
  pattern: RegExp,
): Promise<ServerSearchResult> {
  let connection: McpConnection | null = null;
  try {
    const serverConfig = getServerConfig(config, serverName);
    connection = await getConnection(serverName, serverConfig);

    const tools = await connection.listTools();
    const results: SearchResult[] = [];

    for (const tool of tools) {
      // Match against tool name only (not server name or description)
      if (pattern.test(tool.name)) {
        results.push({ server: serverName, tool });
      }
    }

    debug(`${serverName}: found ${results.length} matches`);
    return { serverName, results };
  } catch (error) {
    const errorMsg = (error as Error).message;
    debug(`${serverName}: connection failed - ${errorMsg}`);
    return { serverName, results: [], error: errorMsg };
  } finally {
    if (connection) {
      await safeClose(connection.close);
    }
  }
}

/**
 * Execute the grep command
 */
export async function grepCommand(options: GrepOptions): Promise<void> {
  let config: McpServersConfig;

  try {
    config = await loadConfig(options.configPath);
  } catch (error) {
    console.error((error as Error).message);
    process.exit(ErrorCode.CLIENT_ERROR);
  }

  const pattern = globToRegex(options.pattern);
  const serverNames = listServerNames(config);

  if (serverNames.length === 0) {
    console.error(
      'Warning: No servers configured. Add servers to mcp_servers.json',
    );
    return;
  }

  const concurrencyLimit = getConcurrencyLimit();

  debug(
    `Searching ${serverNames.length} servers for pattern "${options.pattern}" (concurrency: ${concurrencyLimit})`,
  );

  // Process servers in parallel with concurrency limit
  const serverResults = await processWithConcurrency(
    serverNames,
    (serverName) => searchServerTools(serverName, config, pattern),
    concurrencyLimit,
  );

  const allResults: SearchResult[] = [];
  const failedServers: string[] = [];

  for (const result of serverResults) {
    allResults.push(...result.results);
    if (result.error) {
      failedServers.push(result.serverName);
    }
  }

  // Show failed servers warning
  if (failedServers.length > 0) {
    console.error(
      `Warning: ${failedServers.length} server(s) failed to connect: ${failedServers.join(', ')}`,
    );
  }

  if (allResults.length === 0) {
    console.log(`No tools found matching "${options.pattern}"`);
    console.log(`  Tip: Pattern matches tool names only (not server names)`);
    console.log(`  Tip: Use '*' for wildcards, e.g. '*file*' or 'read_*'`);
    console.log(`  Tip: Run 'mcp-cli' to list all available tools`);
    return;
  }

  // Human-readable output
  console.log(formatSearchResults(allResults, options.withDescriptions));
}

File: src/commands/grep.rs
use crate::config::debug;
use crate::errors::CliError;
use std::sync::Arc;
use tokio::sync::Semaphore;

pub struct GrepOptions {
    pub pattern: String,
    pub with_descriptions: bool,
    pub config_path: Option<String>,
}

pub struct SearchResult {
    pub server: String,
    pub tool: crate::output::ToolInfo,
}

struct ServerSearchResult {
    server_name: String,
    results: Vec<SearchResult>,
    error: Option<String>,
}

pub fn glob_to_regex(pattern: &str) -> Result<regex::Regex, regex::Error> {
    let mut escaped = String::new();
    let chars: Vec<char> = pattern.chars().collect();
    let mut i = 0;

    while i < chars.len() {
        let char = chars[i];

        if char == '*' && i + 1 < chars.len() && chars[i + 1] == '*' {
            escaped.push_str(".*");
            i += 2;
            while i < chars.len() && chars[i] == '*' {
                i += 1;
            }
        } else if char == '*' {
            escaped.push_str("[^/]*");
            i += 1;
        } else if char == '?' {
            escaped.push_str("[^/]");
            i += 1;
        } else if "[.+^${}()|\\]".contains(char) {
            escaped.push('\\');
            escaped.push(char);
            i += 1;
        } else {
            escaped.push(char);
            i += 1;
        }
    }

    regex::RegexBuilder::new(&format!("^{}$", escaped))
        .case_insensitive(true)
        .build()
}

async fn search_server_tools(
    server_name: &str,
    config: &crate::config::McpServersConfig,
    pattern: &regex::Regex,
) -> ServerSearchResult {
    let server_config = match crate::config::get_server_config(config, server_name) {
        Ok(sc) => sc,
        Err(e) => {
            return ServerSearchResult {
                server_name: server_name.to_string(),
                results: Vec::new(),
                error: Some(e.message),
            };
        }
    };

    match crate::client::get_connection(server_name, &server_config).await {
        Ok(connection) => match connection.list_tools().await {
            Ok(tools) => {
                let mut results = Vec::new();
                for tool in tools {
                    if pattern.is_match(&tool.name) {
                        results.push(SearchResult {
                            server: server_name.to_string(),
                            tool,
                        });
                    }
                }
                let _ = connection.close().await;
                debug(&format!("{}: found {} matches", server_name, results.len()));
                ServerSearchResult {
                    server_name: server_name.to_string(),
                    results,
                    error: None,
                }
            }
            Err(e) => {
                let _ = connection.close().await;
                debug(&format!(
                    "{}: tools listing failed - {}",
                    server_name, e.message
                ));
                ServerSearchResult {
                    server_name: server_name.to_string(),
                    results: Vec::new(),
                    error: Some(e.message),
                }
            }
        },
        Err(e) => {
            debug(&format!(
                "{}: connection failed - {}",
                server_name, e.message
            ));
            ServerSearchResult {
                server_name: server_name.to_string(),
                results: Vec::new(),
                error: Some(e.message),
            }
        }
    }
}

pub async fn grep_command(options: GrepOptions) -> Result<(), CliError> {
    let config = crate::config::load_config(options.config_path.as_deref())?;

    let pattern = glob_to_regex(&options.pattern).map_err(|e| crate::errors::CliError {
        code: crate::errors::ErrorCode::ClientError,
        error_type: "INVALID_PATTERN".to_string(),
        message: format!("Invalid glob pattern: {}", e),
        details: None,
        suggestion: None,
    })?;

    let server_names = crate::config::list_server_names(&config);

    if server_names.is_empty() {
        eprintln!("Warning: No servers configured. Add servers to mcp_servers.json");
        return Ok(());
    }

    let concurrency_limit = crate::config::get_concurrency_limit();
    debug(&format!(
        "Searching {} servers for pattern \"{}\" (concurrency: {})",
        server_names.len(),
        options.pattern,
        concurrency_limit
    ));

    let semaphore = Arc::new(Semaphore::new(concurrency_limit));
    let mut tasks = Vec::new();

    for name in server_names {
        let name = name.clone();
        let config = config.clone();
        let pattern_clone = pattern.clone();
        let sem = Arc::clone(&semaphore);

        tasks.push(tokio::spawn(async move {
            let _permit = sem.acquire().await.unwrap();
            search_server_tools(&name, &config, &pattern_clone).await
        }));
    }

    let mut all_results = Vec::new();
    let mut failed_servers = Vec::new();

    for task in tasks {
        if let Ok(res) = task.await {
            all_results.extend(res.results);
            if let Some(_err) = res.error {
                failed_servers.push(res.server_name);
            }
        }
    }

    if !failed_servers.is_empty() {
        eprintln!(
            "Warning: {} server(s) failed to connect: {}",
            failed_servers.len(),
            failed_servers.join(", ")
        );
    }

    if all_results.is_empty() {
        println!("No tools found matching \"{}\"", options.pattern);
        println!("  Tip: Pattern matches tool names only (not server names)");
        println!("  Tip: Use '*' for wildcards, e.g. '*file*' or 'read_*'");
        println!("  Tip: Run 'mcp-cli' to list all available tools");
        return Ok(());
    }

    let display_results: Vec<crate::output::SearchResult> = all_results
        .into_iter()
        .map(|r| crate::output::SearchResult {
            server: r.server,
            tool: r.tool,
        })
        .collect();

    println!(
        "{}",
        crate::output::format_search_results(&display_results, options.with_descriptions)
    );

    Ok(())
}
